src/libraries/System.Text.Json/System.Text.Json.SourceGeneration/JsonSourceGenerator.cs

```diff
@@ -16,47 +18,105 @@ namespace System.Text.Json.SourceGeneration
+
+            MetadataLoadContext mlc = new MetadataLoadContext(context.Compilation);
+
+            SemanticModel semanticModel;
+            INamedTypeSymbol namedTypeSymbol;
+            ITypeSymbol typeSymbol;
+
+            // Map type name to type objects.
+            foreach (KeyValuePair<string, TypeDeclarationSyntax> entry in receiver.InternalClassTypeDict)
+            {
+                TypeDeclarationSyntax tds = entry.Value;
+                semanticModel = context.Compilation.GetSemanticModel(tds.SyntaxTree);
+                namedTypeSymbol = (INamedTypeSymbol)semanticModel.GetDeclaredSymbol(tds);
+                Type convertedType = new TypeWrapper(namedTypeSymbol, mlc);
+                foundTypes[entry.Key] = convertedType;
+            }
+            foreach (KeyValuePair<string, IdentifierNameSyntax> entry in receiver.ExternalClassTypeDict)
+            {
+                IdentifierNameSyntax ins = entry.Value;
+                semanticModel = context.Compilation.GetSemanticModel(ins.SyntaxTree);
+                typeSymbol = context.Compilation.GetSemanticModel(ins.SyntaxTree).GetTypeInfo(ins).ConvertedType;
+                Type convertedType = new TypeWrapper(typeSymbol, mlc);
+                foundTypes[entry.Key] = convertedType;
+            }
+
+            // Create sources for all found types.
+            foreach (KeyValuePair<string, Type> entry in foundTypes)
+            {
+                context.AddSource($"{entry.Key}ClassInfo", SourceText.From($@"
 using System;
-namespace HelloWorldGenerated
-{
-    public static class HelloWorld
-    {
-        public static string SayHello() 
-        {
-            return ""Hello"";
-");
 
-            sourceBuilder.Append(@"
-        }
-    }
-}");
+namespace HelloWorldGenerated
+{{
+    public class {entry.Key}ClassInfo
+    {{
+        public {entry.Key}ClassInfo() {{ }}
 
-            context.AddSource("helloWorldGenerated", SourceText.From(sourceBuilder.ToString(), Encoding.UTF8));
+        public string TestMethod()
+        {{
+            return ""{entry.Key}"";
+        }}
+    }}
+}}
+", Encoding.UTF8));
+            }
         }
 
         public void Initialize(InitializationContext context)
         {
             context.RegisterForSyntaxNotifications(() => new JsonSerializableSyntaxReceiver());
         }
 
-        // Temporary function for now that reads all types. Should search types with attribute JsonSerializable.
-        internal class JsonSerializableSyntaxReceiver : ISyntaxReceiver
+        public class JsonSerializableSyntaxReceiver : ISyntaxReceiver
         {
-            public List<TypeDeclarationSyntax> GeneratorInputTypes = new List<TypeDeclarationSyntax>();
+            public Dictionary<string, IdentifierNameSyntax> ExternalClassTypeDict = new Dictionary<string, IdentifierNameSyntax>();
+            public Dictionary<string, TypeDeclarationSyntax> InternalClassTypeDict = new Dictionary<string, TypeDeclarationSyntax>();
```

Comment:
Do the keys for these dictionaries account for type name collisions between namespaces/assemblies?

Reply:
Also, why dictionaries and not something like `List<T>`?

Reply:
I was going to use dictionaries to be able to detect name collisions since we would use the found strings to directly generate the code. The name collision between namespaces/assemblies are not accounted for yet. Will work on it.